File: code/rabbitmq/csharp/Broker.cs
using System.Text;
using RabbitMQ.Client;

namespace Learn.Rabbit;

// What every example needs: a connection to the course broker, and text in and out of message bodies.
static class Broker
{
    public static readonly string Host = Environment.GetEnvironmentVariable("RABBITMQ_HOST") ?? "localhost";

    // One connection per application, opened once and kept: it is a TCP connection with a handshake.
    // The name shows up in the management UI and in rabbitmqctl list_connections.
    public static Task<IConnection> ConnectAsync(string name) =>
        new ConnectionFactory { HostName = Host, ClientProvidedName = name }.CreateConnectionAsync();

    public static byte[] Bytes(string text) => Encoding.UTF8.GetBytes(text);

    public static string Text(ReadOnlyMemory<byte> body) => Encoding.UTF8.GetString(body.Span);

    // Server properties are AMQP tables: strings arrive as byte arrays.
    public static string Text(object? value) => value is byte[] bytes ? Encoding.UTF8.GetString(bytes) : $"{value}";

    // Starts each example from a clean slate, so that its output is the same on every run.
    public static async Task DeleteAsync(IChannel channel, string[] exchanges, string[] queues)
    {
        foreach (var exchange in exchanges)
        {
            await channel.ExchangeDeleteAsync(exchange);
        }
        foreach (var queue in queues)
        {
            await channel.QueueDeleteAsync(queue, ifUnused: false, ifEmpty: false);
        }
    }

    // Empties a queue with basic.get and prints the bodies it held, in queue order.
    public static async Task PrintQueueAsync(IChannel channel, string queue)
    {
        var lines = new List<string>();
        while (await channel.BasicGetAsync(queue, autoAck: true) is { } message)
        {
            lines.Add(Text(message.Body));
        }
        Console.WriteLine($"{queue}: {(lines.Count == 0 ? "(empty)" : string.Join(" | ", lines))}");
    }

    // Waits for a condition set by a consumer callback, and fails the example instead of hanging.
    public static async Task WaitAsync(Task task, string what)
    {
        if (await Task.WhenAny(task, Task.Delay(TimeSpan.FromSeconds(10))) != task)
        {
            throw new TimeoutException($"timed out waiting for {what}");
        }
        await task;
    }
}

File: code/rabbitmq/java/client/src/main/java/dev/learn/rabbitmq/Broker.java
package dev.learn.rabbitmq;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.GetResponse;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/** What every example needs: a connection to the course broker, and text in and out of message bodies. */
final class Broker {

    private Broker() {
    }

    static final String HOST = System.getenv().getOrDefault("RABBITMQ_HOST", "localhost");

    // One connection per application, opened once and kept. The name shows up in the management UI.
    static Connection connect(String name) throws IOException, TimeoutException {
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(HOST);
        return factory.newConnection(name);
    }

    static byte[] bytes(String text) {
        return text.getBytes(StandardCharsets.UTF_8);
    }

    static String text(byte[] body) {
        return new String(body, StandardCharsets.UTF_8);
    }

    // Empties a queue with basic.get and prints the bodies it held, in queue order.
    static void printQueue(Channel channel, String queue) throws IOException {
        List<String> lines = new ArrayList<>();
        GetResponse message;
        while ((message = channel.basicGet(queue, true)) != null) {
            lines.add(text(message.getBody()));
        }
        System.out.println(queue + ": " + (lines.isEmpty() ? "(empty)" : String.join(" | ", lines)));
    }

    // Waits for a consumer callback, and fails the example instead of hanging.
    static void await(CountDownLatch latch, String what) throws InterruptedException {
        if (!latch.await(10, TimeUnit.SECONDS)) {
            throw new IllegalStateException("timed out waiting for " + what);
        }
    }
}
